### PR TITLE
Second try to vectorize sample_posterior_predictive.

### PR DESCRIPTION
**Pull request by rpgoldman**
Not ready for prime time; sharing for interested parties.

Replacement for #3548

### CODE REVIEW

File: pymc3/backends/ndarray.py
Comment (lucianopaz) [inline pymc3/backends/ndarray.py]:
```suggestion
def point_list_to_multitrace(point_list: List[Dict[str, np.ndarray]], model: Optional[Model]=None) -> MultiTrace:
```

File: pymc3/tests/test_sampling.py
Comment (lucianopaz) [inline pymc3/tests/test_sampling.py:355]: Why was this change necessary?

File: pymc3/tests/test_sampling.py
Comment (lucianopaz) [inline pymc3/tests/test_sampling.py]: The `size` argument lets you draw more than one sample per point in the trace. In my opinion, this is necessary when you want to draw posterior predictives samples using a single point estimate, like the MAP or MLE.

**Review by lucianopaz — COMMENTED** [review]
Thanks @rpgoldman. I left a couple of minor comments written down.

I'm still very hesitant about this PR because of the amount of code duplication and added codebase complexity that it brings with `posterior_predictive.py`. Ideally, we would want to just make a minor change to `sample_posterior_predictive`, leaving most of the `draw_values`, `_draw_value`, etc parts as they are. The ideal change would be to use the current `sample_posterior_predictive` but instead of passing each `point` in the trace in a `for` loop, we would pass a `point={trace.get_values(rv_name) for rv_name in trace.varnames}`. This would pass the entire trace all at once and `draw_values` should have to be able to deal with everything else. What prevents us from doing this now are the multivariate distributions, that are not properly vectorized. I propose that we do something different like:

1. We create a small function in `sampling.py` (lets call it `fast_sample_posterior_predictive`)
2. This function would be almost a complete copy of the original `sample_posterior_predictive`. The only change would be that we wouldn't loop over the points in the trace one at a time, but create a dictionary `{rv_name: trace.get_values(rv_name) for rv_name in trace.varnames}` and pass that as the point.
3. Models with multivariate distributions would fail at this until we finish vectorizing the multivariates properly.

File: pymc3/tests/test_sampling.py
Comment (rpgoldman) [reply pymc3/tests/test_sampling.py:355]: This is necessary for now because I rely on the trace argument to have the right length for sampling.
Since the trace is always a multi-trace, I demand that the `samples` argument be a multiple of the number of chains in the `MultiTrace`. If I knew how to take a `MultiTrace` and smash all of the component chains together into a single chain, this would not be necessary.
If you know how to do that, that would be great.

File: pymc3/tests/test_sampling.py
Comment (rpgoldman) [reply pymc3/tests/test_sampling.py]: Maybe that's true, but that requires handling *both* `size` *and* `samples`, and the necessary infrastructure in terms of `draw_value`, and `random` does not exist for this, as far as I can tell.
Also, the use of `size` somehow changes the shape of the returned traces, which is a further complication, an added circle in shape hell.
I'd be happy to restore this if I could see a way to do it. But I can't right now.

**Comment by rpgoldman** [comment]
> Thanks @rpgoldman. I left a couple of minor comments written down.
> 
> I'm still very hesitant about this PR because of the amount of code duplication and added codebase complexity that it brings with `posterior_predictive.py`. 

I agree about that, but I want to do this in a "hill-climbing" way: that is, I don't want to break any *other* parts of PyMC3 while I am getting this fixed. So I would rather have duplicate code, at least for now, than break prior predictive sampling, or other such code, on my branch.

So I am generally in agreement about your remarks below, but I'm not sure how to get from here to there, especially given the problems with multivariate distributions.

> Ideally, we would want to just make a minor change to `sample_posterior_predictive`, leaving most of the `draw_values`, `_draw_value`, etc parts as they are. The ideal change would be to use the current `sample_posterior_predictive` but instead of passing each `point` in the trace in a `for` loop, we would pass a `point={trace.get_values(rv_name) for rv_name in trace.varnames}`. This would pass the entire trace all at once and `draw_values` should have to be able to deal with everything else. What prevents us from doing this now are the multivariate distributions, that are not properly vectorized.

I think it's more than just that -- there are also at least two further complications, IIUC: 

1. `sample_posterior_predictive` uses *both* `size` *and* `samples`, and as far as I can tell the other APIs don't support both.  And what I mean by that is that generate samples and random methods don't support them, which would impose substantial pain.  Why not remove it, as I have done, and just sample repeatedly as needed?  I.e., control entirely through `samples` instead? Especially since introducing `size` changes the shape of the output?

2. I don't understand to what extent the theano code relies on having the right input shapes. That's why you will see that I have been slicing the input trace so that its length matches the number of samples (and this accounts for my demanding that the `samples` argument be a multiple of the number of chains in the trace.

> I propose that we do something different like:
> 
>     1. We create a small function in `sampling.py` (lets call it `fast_sample_posterior_predictive`)
> 
>     2. This function would be almost a complete copy of the original `sample_posterior_predictive`. The only change would be that we wouldn't loop over the points in the trace one at a time, but create a dictionary `{rv_name: trace.get_values(rv_name) for rv_name in trace.varnames}` and pass that as the point.

I think I needed to do some things in the guts of `draw_value` to make it work with vectorization, including changing the code that makes inferences about shape, but those are probably limited.  If necessary, maybe we could have a `trace` argument in addition to a `point` argument and branch on which is passed.

I like the idea of having a dictionary instead of a `MultiTrace`, and I could more towards that, once all the tests are passing.  Especially since it would remove the multiple chain issues and the issues from building the `MultiTrace` by hand.
> 
>     3. Models with multivariate distributions would fail at this until we finish vectorizing the multivariates properly.

This would be a new limitation, right?  I.e., `sample_posterior_predictive` works on multivariates now, without vectorization, but won't when it's vectorized instead of point at a time?

**Comment by rpgoldman** [comment]
@lucianopaz Maybe as a step towards reducing the duplication of code, we could try to identify the differences between `posterior_predictive_draw_values` and `draw_values` with an eye to bringing them back together?
I guess this would be easier if we could also vectorize `sample_prior_predictive`. But if we don't do that -- and if we cannot vectorize sampling from multivariate distributions, I don't see how we can avoid code duplication.
One thing that's difficult is that I haven't found a good substitute for my use of `contextvars` to override `draw_values` with `posterior_predictive_draw_values`.  I hadn't realized that this was 3.7 specific.  I tried replacing the context variable with a global variable, but it didn't work -- I'm not sure if that's due to a threading issue or what.

File: pymc3/tests/test_sampling.py
Comment (lucianopaz) [reply pymc3/tests/test_sampling.py:355]: The `get_values` method combines all chains together for a single variable. That's why `{rv_name: trace.get_values(rv_name) for rv_name in trace.varnames}` would give you a single dictionary, the keys would be the variable names and the values would be all the samples from every chain in the trace. Wouldn't this dictionary be tracelike?

Comment (rpgoldman) [reply pymc3/tests/test_sampling.py:355]: > The `get_values` method combines all chains together for a single variable. That's why `{rv_name: trace.get_values(rv_name) for rv_name in trace.varnames}` would give you a single dictionary, the keys would be the variable names and the values would be all the samples from every chain in the trace. Wouldn't this dictionary be tracelike?

Oh, excellent! I will look at that. The only remaining issue would be dealing with the absence of `trace.varnames`, and the fact that it doesn't support a `points` but that would be easy to deal with building on UserDict.  I'll see about using `get_values` and removing the limitation on allowable `samples` values.  That will be an easier canonical representation than the `MultiTrace`.

BTW, I am down to 2 tests failing, and of those 2, one would be fixed by this change and one is a multivariate issue which AFAICT simply isn't fixable by me.

Speaking of which, is there some test which would detect any and all multivariate distributions so that I could make a new `fast_posterior_predictive` yield a *useful* error message, instead of crashing in the middle of `random`?

File: requirements-dev.txt
Comment (fonnesbeck) [inline requirements-dev.txt]: This is already in the requirements.txt changes below

File: pymc3/smc/smc_utils.py
Comment (fonnesbeck) [inline pymc3/smc/smc_utils.py]: Is this a rebase problem?

File: .pycheckers
Comment (fonnesbeck) [inline .pycheckers]: newline

File: pymc3/smc/smc_utils.py
Comment (rpgoldman) [reply pymc3/smc/smc_utils.py]: Yes, I believe so -- I think that when git ran `diff` it decided, mistakenly, that I had added this whole file.

Thanks for catching that, deleting.

File: .pycheckers
Comment (rpgoldman) [reply .pycheckers]: This file simply shouldn't be there -- it's a configuration file used by emacs, and should be ignored. I'll fix that, too.

**Comment by rpgoldman** [comment]
I believe that this is ready for merging now.  Ordinarily, I would recommend squash merging, but in this case the history is probably of interest, so I recommend a normal merge.  I _did_ edit the history to make it clearer.

File: pymc3/distributions/.dir-locals.el
Comment (twiecki) [inline pymc3/distributions/.dir-locals.el]: Accidental commit.

File: pymc3/distributions/__init__.py
Comment (twiecki) [inline pymc3/distributions/__init__.py:174]:
```suggestion
           'fast_sample_posterior_predictive',
```

File: pymc3/distributions/posterior_predictive.py
Comment (twiecki) [inline pymc3/distributions/posterior_predictive.py]: Should this be here?

File: pymc3/distributions/posterior_predictive.py
Comment (twiecki) [inline pymc3/distributions/posterior_predictive.py]:
```suggestion
    def draw_value(self, param, trace: Optional[_TraceDict]=None, givens=None):
```

File: pymc3/distributions/.dir-locals.el
Comment (rpgoldman) [reply pymc3/distributions/.dir-locals.el]: Thanks for catching that.  Fixed in rebase, so it's removed from the history (wait for force push).

File: pymc3/distributions/posterior_predictive.py
Comment (rpgoldman) [reply pymc3/distributions/posterior_predictive.py]: Removed, thanks.

File: pymc3/distributions/posterior_predictive.py
Comment (rpgoldman) [reply pymc3/distributions/posterior_predictive.py]: Thanks for catching that: not sure why pylint didn't complain.

**Comment by rpgoldman** [comment]
@twiecki I have addressed all your comments, for which many thanks.  New version pushed and tests are running.  Also rebased on master, for a clean merge.

**Comment by rpgoldman** [comment]
This test shows up as failing, but it's a timeout error -- there are no actually failed tests.

File: pymc3/pytestdebug.log
Comment (lucianopaz) [inline pymc3/pytestdebug.log]: This log file shouldn't be here. Maybe you could add a rule to .gitignore

File: pymc3/tests/test_data_container.py
Comment (lucianopaz) [inline pymc3/tests/test_data_container.py:52]: Why does this have a different shape than the one above?

**Review by lucianopaz — CHANGES_REQUESTED** [review]
I know that you've worked on this for a very long time and I haven't reviewed it in quite a while. My main critique was and still is code complexity. I honestly can't devout much time to pymc3, and I'm focusing the time I have on pymc4 instead, so I don't want to block this PR from being merged. I've left two comments above and once you reply to those, I'll merge this into master, but forward sampling bug fixes from the time of the merge onwards will have to be applied to two independent forward sampling branches

File: pymc3/pytestdebug.log
Comment (rpgoldman) [reply pymc3/pytestdebug.log]: I will remove it from the history.  There's already a rule for it in .gitignore.

File: pymc3/tests/test_data_container.py
Comment (rpgoldman) [reply pymc3/tests/test_data_container.py:52]: This happens because the shape of the predictor variable (`x_shared`) has changed.  It was originally a vector of size 100 (line 36), but on line 45, we update it to be of size 200 (see line 33 for the new value).  This changes the size of the prior predictive trace.

This was always true, I just added a check to verify it.

**Comment by rpgoldman** [comment]
> I know that you've worked on this for a very long time and I haven't reviewed it in quite a while. My main critique was and still is code complexity. I honestly can't devout much time to pymc3, and I'm focusing the time I have on pymc4 instead, so I don't want to block this PR from being merged. I've left two comments above and once you reply to those, I'll merge this into master, but forward sampling bug fixes from the time of the merge onwards will have to be applied to two independent forward sampling branches

I understand the problem, but I am not sure what to do about it.  There is still the possibility that the two versions of posterior predictive sampling are incomparable, in the sense that they may each have slightly different bug cases.  I don't know what to do about this beyond checking *both* versions each time we introduce a bug fix.

The only thing I can suggest, given everyone's time constraints, is that you try to ensure that forward/predictive sampling not be an afterthought in PyMC4 the way it is in PyMC3.

Particularly of concern: make sure that transformations work properly in both directions, not only in one direction as in PyMC3, and that mixture models can be forward sampled.

End of editorial/sermon! 😉

File: pymc3/pytestdebug.log
Comment (rpgoldman) [reply pymc3/pytestdebug.log]: Removed it from the history.  There's a rule for it in .gitignore.   I must have accidentally committed it in the blob immediately before the blob where I committed the rule in .gitignore.

File: pymc3/tests/test_data_container.py
Comment (rpgoldman) [reply pymc3/tests/test_data_container.py:52]: It has a different shape because of the change to the predictor variable.  When we reset the `Data` item, we change the shape of the predictor, and so we change the shape of the observed variable.

This is not a change in the function of the code -- I just tested it slightly more thoroughly to verify the prior predictive sampling was working properly (since posterior predictive sampling did not, and does not -- when you change the shape of the predictors, in general, you must get a new model and may need to edit the posterior trace.).

**Comment by lucianopaz** [comment]
Great work @rpgoldman!

**Pull request merged**
